### PR TITLE
WIP - Allow fields to be selected with "selector" aggregate functions

### PR DESCRIPTION
**Pull request by corylanou**
Heavily WIP, needs refactoring, and a lot of tests.

### CODE REVIEW

File: tsdb/functions.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 // These are used by the mapFunctions in this file
 type Iterator interface {
 	Next() (time int64, value interface{})
+	Fields() map[string]interface{}
```
Comment (nathanielc) [inline tsdb/functions.go]: Am I correct in assuming that the Fields() return value is different after each call to Next?

File: tsdb/functions.go [hunk 2]
```diff
@@ -60,15 +61,21 @@ func initializeMapFunc(c *influxql.Call) (mapFunc, error) {
 	case "distinct":
 		return MapDistinct, nil
 	case "sum":
-		return MapSum, nil
+		return func(itr Iterator) interface{} {
```
Comment (nathanielc) [inline tsdb/functions.go]: Would it be possible to abstract the parsing of arguments from the `Call` object. 

For example `MapSum` sums the values of a field. It needs the call object to know which field to sum.

Could we make the `MapSum` signature 
```func MapSum(itr Iterator, field string) interface{}```
And then extract the field from the `Call` object here in this anon func. And just pass an empty field if it doesn't exist.

```
return func(itr Iterator) interface{} {
        var  field string
        if fields := c.Fields(); len(fields) > 0 {
                field = fields[0]
        }
        return MapSum(itr, field)
}, nil
```

A quick glance at the uses of `Call` in the functions shows that they are used in a similar manner and that the logic is not dependent on the data in the itr.

**Comment by nathanielc** [comment]
Overall this will actually be a very positive change for my MR needs as I was already planning on preserving the Fields and Tags through the framework.

**Comment by corylanou** [comment]
closing in favor of #4202

File: tsdb/functions.go (hunk 2 above)
Comment (corylanou) [inline tsdb/functions.go]: Yeah, that should be a simple enough refactor.

File: tsdb/functions.go (hunk 1 above)
Comment (corylanou) [inline tsdb/functions.go]: correct.

**Pull request closed without merging**